apps/one-drive/src/inngest/functions/data-protection/sync-items.ts

```diff
@@ -0,0 +1,302 @@
+import type { MicrosoftDriveItemPermissions } from '../../../connectors/share-point/permissions';
+import { getItemPermissions } from '../../../connectors/share-point/permissions';
+
+export type ItemsWithPermisions = {
+  item: MicrosoftDriveItem;
+  permissions: MicrosoftDriveItemPermissions[];
+};
+
+export const parseItems = (
+  inputItems: MicrosoftDriveItem[]
+): {
+  folders: MicrosoftDriveItem[];
+  items: MicrosoftDriveItem[];
+} => {
+  const folders: MicrosoftDriveItem[] = [];
+  const items: MicrosoftDriveItem[] = [];
+
+  inputItems.forEach((item) => {
+    if (item.folder) {
+      folders.push(item);
+    } else {
+      items.push(item);
+    }
+  });
+
+  return {
+    folders,
+    items,
+  };
+};
+
+export const getCkunkedArray = <T>(array: T[], batchSize: number): T[][] => {
+  const chunks: T[][] = [];
+  for (let i = 0; i < array.length; i += batchSize) {
+    chunks.push(array.slice(i, i + Number(batchSize)));
+  }
+  return chunks;
+};
+
+export const formatPermissions = (
+  permission: MicrosoftDriveItemPermissions
+): DataProtectionObjectPermission[] | [] => {
+  if (permission.grantedToV2?.user) {
+    return [
+      {
+        id: permission.id,
+        type: 'user',
+        displayName: permission.grantedToV2.user.displayName,
+        userId: permission.grantedToV2.user.id,
+        email: permission.grantedToV2.user.email,
+        metadata: {
+          permissionId: permission.id,
+          roles: permission.roles,
+        },
+      },
+    ];
+  } else if (permission.grantedToIdentitiesV2) {
+    const formattedPermissions: DataProtectionObjectPermission[] = [];
+
+    for (const p of permission.grantedToIdentitiesV2) {
+      if (p.user) {
+        formattedPermissions.push({
+          id: `${permission.id}${env.ID_SEPARATOR}${p.user.id}`,
+          type: 'user',
+          displayName: p.user.displayName,
+          userId: p.user.id,
+          email: p.user.email,
+          metadata: {
+            permissionId: permission.id,
+            roles: permission.roles,
+          },
+        });
+      }
+    }
+
+    return formattedPermissions;
+  }
+  return [];
+};
+
+export const parseDataProtetionItems = (
+  itemsWithPermisions: ItemsWithPermisions[]
+): DataProtectionObject[] => {
+  const dataProtection: DataProtectionObject[] = [];
+
+  for (const { item, permissions } of itemsWithPermisions) {
+    if (item.createdBy.user.id) {
+      const validPermissions: MicrosoftDriveItemPermissions[] = permissions.filter((p) => {
+        if (p.link && p.link.scope === 'users') return true;
+        else if (p.grantedToV2?.user) return true;
+        return false;
+      });
+
+      if (validPermissions.length) {
+        const dataProtectionItem = {
+          id: item.id,
+          name: item.name,
+          ownerId: item.createdBy.user.id,
+          url: item.webUrl,
+          permissions: validPermissions.map((p) => formatPermissions(p)).flat(),

```

Comment:
```suggestion
          permissions: validPermissions.map(formatPermissions),

```

`formatPermissions` already returns a flat array, isn't it ?

Reply:
`formatPermissions` returns flat array, but then we use map and get array of arrays, so we need to flat this here.
Here we can get one permission or array of permissions from each `(p)` object, so i thought this would be a simplier option.